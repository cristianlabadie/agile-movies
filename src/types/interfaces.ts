export interface ResponseError {
  message: string
  statusCode: number
}
export interface ResponseLogin {
  data: Data
}

export interface Data {
  payload: Payload
  user: User
}

export interface Payload {
  refresh_token: string
  token: string
  type: string
}

export interface User {
  email: string
  firstName: string
  lastName: string
}

export interface ResponseJWTError {
  error: string
  message: string
  statusCode: number
}

export interface ResponseUserMe {
  data: Data
}

export interface UseInfo {
  createdAt: Date
  email: string
  firstName: string
  id: number
  lastName: string
  password: null
  username: string
}

export interface ResponseMovies {
  imageBaseUrl: string
  data: Movie[]
}

export interface Movie {
  adult: boolean
  backdrop_path: string
  genre_ids: number[]
  id: number
  original_language: OriginalLanguage
  original_title: string
  overview: string
  popularity: number
  poster_path: string
  release_date: Date
  title: string
  video: boolean
  vote_average: number
  vote_count: number
}

export enum OriginalLanguage {
  En = 'en',
  Ja = 'ja',
  No = 'no',
  Tl = 'tl'
}

export interface ResponseActors {
  imageBaseUrl: string
  data: Actor[]
}

export interface Actor {
  adult: boolean
  gender: number
  id: number
  known_for_department: KnownForDepartment
  name: string
  original_name: string
  popularity: number
  profile_path: null | string
  cast_id: number
  character: string
  credit_id: string
  order: number
}

export enum KnownForDepartment {
  Acting = 'Acting',
  Crew = 'Crew'
}
