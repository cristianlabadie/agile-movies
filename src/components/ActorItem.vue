<template>
  <div class="min-w-32 select-none flex flex-col gap-1 dark:text-white items-center">
    <div v-if="loadingImg">
      <div class="animate-pulse">
        <div class="bg-gray-300 dark:bg-gray-600 w-32 h-48 rounded-md"></div>
      </div>
    </div>

    <img
      v-else-if="!loadingImg && errorOnLoad"
      src="https://via.placeholder.com/300x450/000000/FFFFFF/?text=No%20Image"
      :alt="actor.name"
      class="w-full h-48 rounded-md"
    />

    <img
      v-else-if="!loadingImg && actor.profile_path"
      :src="imgActor.src"
      :alt="actor.name"
      class="w-auto h-48 rounded-md"
      id="actor-img"
    />

    <div class="w-full">
      <!-- Actor info -->
      <h3 class="text-xs text-pretty line-clamp-1" id="character">
        {{ actor.character }}
      </h3>

      <small class="text-gray-400 truncate" id="name">
        {{ actor.name }}
      </small>
    </div>
  </div>
</template>

<script setup lang="ts">
import type { Actor } from '@/types/interfaces'
import { ref, type PropType, onMounted } from 'vue'

const props = defineProps({
  actor: {
    type: Object as PropType<Actor>,
    required: true
  },
  baseUrl: {
    type: String,
    required: true
  }
})

const imgActor = new Image()
const loadingImg = ref(true)
const errorOnLoad = ref(false)

imgActor.onload = () => {
  loadingImg.value = false
}
imgActor.onerror = () => {
  errorOnLoad.value = false
}

onMounted(() => {
  if (!props.actor.profile_path) {
    loadingImg.value = false
    errorOnLoad.value = true
    imgActor.src = 'https://via.placeholder.com/300x450/000000/FFFFFF/?text=No%20Image'
  } else {
    imgActor.src = props.baseUrl + props.actor.profile_path
  }
})
</script>
