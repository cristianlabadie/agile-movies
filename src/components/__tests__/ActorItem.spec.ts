import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import { createTestingPinia } from '@pinia/testing'
import ActorItem from '../ActorItem.vue'
import { vi } from 'vitest'
import { KnownForDepartment } from '@/types/interfaces'

describe('Pruebas en ActorItem component', () => {
  const actor = {
    adult: false,
    gender: 2,
    id: 117642,
    known_for_department: KnownForDepartment.Acting,
    name: 'Jason Momoa',
    original_name: 'Jason Momoa',
    popularity: 81.044,
    profile_path: '/6dEFBpZH8C8OijsynkSajQT99Pb.jpg',
    cast_id: 35,
    character: 'Arthur Curry / Aquaman',
    credit_id: '629e1e1ea284eb009df9bb9d',
    order: 0
  }

  const wrapper = mount(ActorItem, {
    global: {
      plugins: [
        createTestingPinia({
          createSpy: vi.fn,
          initialState: {
            movies: {
              actors: [actor]
            }
          }
        })
      ]
    },
    props: {
      actor: actor,
      baseUrl: 'https://image.tmdb.org/t/p/w500'
    }
  })

  it('Debe de renderizar el componente correctamente', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('Debe de mostrar el nombre del actor', () => {
    const name = wrapper.find('#name')
    expect(name.text()).toBe(actor.name)
  })

  it('Debe de mostrar el personaje del actor', () => {
    const character = wrapper.find('#character')
    expect(character.text()).toBe(actor.character)
  })
})
