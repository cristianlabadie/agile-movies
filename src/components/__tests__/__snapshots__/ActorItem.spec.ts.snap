// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pruebas en ActorItem component > Debe de renderizar el componente correctamente 1`] = `
"<div class="min-w-32 select-none flex flex-col gap-1 dark:text-white items-center">
  <div>
    <div class="animate-pulse">
      <div class="bg-gray-300 dark:bg-gray-600 w-32 h-48 rounded-md"></div>
    </div>
  </div>
  <div class="w-full">
    <!-- Actor info -->
    <h3 class="text-xs text-pretty line-clamp-1" id="character">Arthur Curry / Aquaman</h3><small class="text-gray-400 truncate" id="name">Jason Momoa</small>
  </div>
</div>"
`;
